### PR TITLE
323 credit issuance

### PR DESCRIPTION
**Pull request by ba1ciu**

### CODE REVIEW

**Review by gjermundgaraba — CHANGES_REQUESTED** [review]

File: chain/proto/empowerchain/plasticcredit/types.proto [hunk 1]
```diff
@@ -53,41 +53,29 @@ message CreditClass {
   ProvenData credit_class_data = 4;  
 }
 
-message CreditBatch {
-  string batch_denom = 1; 
+message CreditCollection {
   uint64 project_id = 2;
```
[the remark this answers is not in the document]
Comment (ba1ciu) [reply chain/proto/empowerchain/plasticcredit/types.proto:57]: No, the "denom" is ID here.

File: chain/x/plasticcredit/msgs.go [hunk 2]
```diff
@@ -75,3 +77,85 @@ func (m *MsgCreateApplicant) GetSigners() []sdk.AccAddress {
 	}
 	return []sdk.AccAddress{admin}
 }
+
+func (m *MsgIssueCredits) ValidateBasic() error {
```
[the remark this answers is not in the document]
Comment (ba1ciu) [reply chain/x/plasticcredit/msgs.go:81]: On it.

File: chain/x/plasticcredit/msgs.go [hunk 3]
```diff
@@ -75,3 +77,85 @@ func (m *MsgCreateApplicant) GetSigners() []sdk.AccAddress {
...
+	_, err := sdk.AccAddressFromBech32(m.Creator)
+	if err != nil {
+		return errors.Wrapf(sdkerrors.ErrInvalidAddress, "invalid creator address (%s)", err)
+	}
+	if m.ProjectId == 0 {
+		return errors.Wrap(ErrInvalidProjectID, "invalid project id")
```
[the remark this answers is not in the document]
Comment (ba1ciu) [reply chain/x/plasticcredit/msgs.go:87]: For msg types validation - I agree, makes sense. Will refactor accordingly.

File: chain/proto/empowerchain/plasticcredit/tx.proto [hunk 4]
```diff
@@ -49,6 +58,59 @@ message MsgCreateIssuerResponse {
...
 }
 
+message MsgIssueCredits {
+  option (cosmos.msg.v1.signer) = "creator";
+
+  string creator = 1 [(cosmos_proto.scalar) = "cosmos.AddressString"];
+
+  uint64 project_id = 2;
+  string denom_suffix = 3;
+  CreditBalance credit_amount = 4;
```
[the remark this answers is not in the document]
Comment (ba1ciu) [reply chain/proto/empowerchain/plasticcredit/tx.proto:68]: It's this way basically to allow for issuing credits that are already retired. Not sure if we need such, but I saw that it's possible on Regen.

File: chain/proto/empowerchain/plasticcredit/types.proto [hunk 5]
```diff
@@ -53,41 +53,29 @@ message CreditClass {
...
-message CreditBalance {
-  string owner = 1; 
-  CreditAmount balance = 2;   
-}
-
-message BatchDenomToCreditBalance {
-  string batch_denom = 1;
-  repeated CreditBalance credit_balances = 2;
+message IDCreditCollection {
+  string denom = 1;
```
[the remark this answers is not in the document]
Comment (ba1ciu) [reply chain/proto/empowerchain/plasticcredit/types.proto:63]: This was my idea of creating different types that also contain keys, but as we now decided to include keys into the store objects, I'm deleting those IDxxx objects.

File: chain/proto/empowerchain/plasticcredit/types.proto (hunk 1 above)
[the remark this answers is not in the document]
Comment (gjermundgaraba) [reply chain/proto/empowerchain/plasticcredit/types.proto:57]: Yeah, I realized eventually :D

File: chain/proto/empowerchain/plasticcredit/tx.proto (hunk 4 above)
[the remark this answers is not in the document]
Comment (gjermundgaraba) [reply chain/proto/empowerchain/plasticcredit/tx.proto:68]: Let's keep it simple until we know we need it.

**Review by gjermundgaraba — APPROVED** [review]

**Pull request merged**